common/hex.c

```diff
@@ -64,3 +66,44 @@ hex_encode (const unsigned char *data,
 	result[o] = 0;
 	return result;
 }
+
+unsigned char *
+hex_decode (const char *hex,
+            size_t *bin_len)
+{
+	size_t i, j;
+	unsigned long val;
+	unsigned char *bin;
+	char hex2[3] = { 0 };
+        size_t hex_len = strlen (hex);
+
+	bin = malloc (hex_len / 2);
+	if (bin == NULL)
+		return NULL;
+
+        for (i = j = 0; i < hex_len;) {
+		if (hex[i] == ':') {
+			++i;
+			continue;
+		}
+
+		if (i + 1 >= hex_len) {
+			free (bin);
+			return NULL;
+		}
+
+		hex2[0] = hex[i++];
+		hex2[1] = hex[i++];
+
+		val = strtoul (hex2, NULL, 16);
```

Comment:
Using `strtoul` for this simple task seems a bit too much for me, as it requires additional setup (`hex2`) and error handling which would unlikely to happen. Maybe you could write expand something like:
```c
/* Calculate bin_len accurately */
return_val_if_fail (SIZE_MAX - 1 >= hex_len, NULL);
return_val_if_fail ((hex_len + 1) % 3 == 0, NULL);

bin_len = (hex_len + 1) / 3;
...
for (i = 0; i < bin_len; i++) {
  int n = 0;
  for (j = 0; j < 2; j++, n <<= 4) {
    int c = hex[i * 3 + j];
    switch (c) {
    case '0': ... case '9':
      n |= c - '0';
      break;
    case 'a': ... case 'f':
      n |= c - 'a';
      break;
    default:
      error;
    }
  c = hex[i * 3 + j];
  if (c != ':' || c != '\0')
    error;
  bin[i] = n;
}
```

Reply:
This way it doesnt accept strings without `:`